src/peergos/server/storage/UserQuotas.java

```diff
@@ -55,6 +55,38 @@ public UserQuotas(JdbcQuotas quotas,
         return Futures.of(getQuota(username));
     }
 
+    /**
+     * Allows Peergos to automatically accept admin-generated requests for more space to be assigned to a specified
+     * user.
+     *
+     * @param adminIdentity Public key hash of the admin user that has sent change quota request
+     * @param signedRequest Request for more space to be assigned to a specified user signed by admin user
+     */
+    @Override
+    public void changeQuota(PublicKeyHash adminIdentity, byte[] signedRequest) {
+        try {
+            Optional<PublicSigningKey> adminPublicKey = dht.getSigningKey(adminIdentity).join();
+            if (!adminPublicKey.isPresent())
+                throw new IllegalStateException("Couldn't retrieve admin key!");
+            CborObject cbor = CborObject.fromByteArray(adminPublicKey.get().unsignMessage(signedRequest));
+            SpaceUsage.SpaceRequest req = QuotaControl.SpaceRequest.fromCbor(cbor);
+
+            if(req.username.equals("*")) {
+                quotas.setAllQuotas(req.bytes);
+            } else {
+                if (!quotas.hasUser(req.username)) {
+                    throw new IllegalStateException("Couldn't lookup given username!");
+                }
+                if (quotas.getQuota(req.username) > req.bytes) {
+                    throw new IllegalStateException("Couldn't reduce quota!");
+                }
+                setQuota(req.username, req.bytes);
```

Comment:
I am just wondering about one thing:

When we try to change quota for single user, we could have a situation that `IllegalStateException("Couldn't reduce quota!")` is raised. But this condition is not checked when we try to update all users (using `*`). Shouldn't  we handle that as well?

Reply:
In the last commit I have moved checking this condition to the database level (JdbcQuotas.java). This is not a perfect solution (it moves some part of the higher-level logic to the lower level), however it allows us to avoid sending separate database updates for each user. On the other hand, if we would like to add for instance excluding admin users, we would have to either do it in this method and send separate updates, or pass the list of admin usernames to JdbcQuotas and modify SQL accordingly. I can revert these last changes so, that the condition is checked in UserQuotas if this feels like a better design.

Reply:
I see, but still I think we have a potential problem. Let's have a scenario that we have 2 users:
1. user `x` has assigned 100 bytes for quota and uses 50 bytes
2. user `y` has assigned 100 bytes for quota and uses 75 bytes

When we try to reduce the number to e.g. 100 for all users ( by using `*` mode), it might lead to unexpected situation.

Might be not for now, but maybe we should get all users and check if for every the new value could be applied?